sycl/doc/extensions/proposed/sycl_ext_oneapi_barrier.asciidoc

```diff
@@ -0,0 +1,260 @@
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2023-2023 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 7 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension defines a `sycl::ext::oneapi::experimental::barrier` class
+aligned with the `std::barrier` class defined in {cpp}20. This extension may
+also be considered a generalization of the existing
+link:../experimental/sycl_ext_oneapi_cuda_async_barrier.asciidoc[sycl_ext_oneapi_cuda_async_barrier]
+extension for non-CUDA backends.
+
+`sycl::ext::oneapi::experimental::barrier` is _not_ a general replacement for
+`sycl::group_barrier`, and is intended for usage by experts seeking to
+synchronize subsets of threads of execution that do not necessarily correspond
+to groups of work-items.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_BARRIER` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== `barrier` class
+
+The `sycl::ext::oneapi::experimental::barrier` class has the same interface as
+the `std::barrier` class from {cpp}20, with an additional `Scope` template
+parameter denoting a memory scope broad enough to contain all threads of
+execution that are able to use the barrier.
```

Comment:
The scope parameter values of global and system might not be possible to implement for every backend. So this specification should provide a way for only a subset of possible scopes to be supported.

Reply:
Good point.  I think that the correct way to do this is to tie things to the existing atomic and USM capabilities, because whether certain barriers are implementable at all will depend upon whether we can fallback to atomics, and whether the barrier object itself can be accessed concurrently by host and device.

Here's my first attempt.  @gmlueck, I'd appreciate your thoughts on this table as well, given what we've discussed on the atomics side.

| Scope | Requirements |
| ------- | --------------- |
| `work_item` | N/A |
| `sub_group` | `info::device::atomic_memory_scope_capabilities` contains `sub_group`. |
| `work_group` | `info::device::atomic_memory_scope_capabilities` contains `work_group`. |
| `device` | `info::device::atomic_memory_scope_capabilities` contains `device`. </br></br> The barrier must be allocated in memory that can be accessed concurrently by all threads of execution using the barrier, and must therefore be allocated in global memory if the barrier is used by work-items in more than one work-group. |
| `system` | `info::device::atomic_memory_scope_capabilities` contains `system`. </br></br> The barrier must be allocated in memory that can be accessed concurrently by all threads of execution using the barrier, as described by `aspect::usm_atomic_host_allocations`, `aspect::usm_atomic_shared_allocations` and `aspect::usm_system_allocations`. |

Reply:
I have a feeling it would be better to provide a new query specifically for the scopes that `barrier` supports.  For example, what if the backend provides native support for `barrier`?  The supported scopes might be different from `atomic_ref` in that case.

Reply:
Added in f72486e7df72.